infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/fsm/WaitSnapshotApplyState.java

```diff
@@ -0,0 +1,190 @@
+                return fsm.getStates().get(LogReplicationStateType.STOPPED);
+            default: {
+                log.warn("Unexpected log replication event {} when in wait snapshot sync apply state.", event.getType());
+                throw new IllegalTransitionException(event.getType(), getType());
+            }
+        }
+    }
+
+    @Override
+    public void onEntry(LogReplicationState from) {
+        log.info("OnEntry :: wait snapshot apply state");
+        this.fsm.getLogReplicationFSMWorkers().submit(this::verifyStatusOfSnapshotSyncApply);
+    }
+
+    private void verifyStatusOfSnapshotSyncApply() {
+        try {
+            log.debug("Verify snapshot sync apply status, sync={}", transitionEventId);
+
+            // Query metadata on remote cluster to verify the status of the snapshot sync apply
+            CompletableFuture<LogReplicationMetadataResponse> metadataResponseCompletableFuture = dataSender.sendMetadataRequest();
+            LogReplicationMetadataResponse metadataResponse = metadataResponseCompletableFuture.get(CorfuLogReplicationRuntime.DEFAULT_TIMEOUT, TimeUnit.MILLISECONDS);
+
+            // If snapshot sync apply phase has been completed on remote cluster, transition to Log Entry Sync
+            // (incremental update replication), otherwise, schedule new query.
+            if (metadataResponse.getLastLogProcessed() == metadataResponse.getSnapshotApplied() &&
+                    metadataResponse.getSnapshotApplied() == baseSnapshotTimestamp) {
+                log.info("Snapshot sync apply is complete appliedTs={}, baseTs={}", metadataResponse.getSnapshotApplied(),
+                        baseSnapshotTimestamp);
+                fsm.input(new LogReplicationEvent(LogReplicationEvent.LogReplicationEventType.SNAPSHOT_APPLY_COMPLETE,
+                        new LogReplicationEventMetadata(transitionEventId, baseSnapshotTimestamp, baseSnapshotTimestamp)));
+            } else {
+                log.debug("Snapshot sync apply is still in progress, appliedTs={}, baseTs={}, sync_id={}", metadataResponse.getSnapshotApplied(),
+                        baseSnapshotTimestamp, transitionEventId);
+                // Schedule a one time action which will verify the snapshot apply status after a given delay
+                this.snapshotSyncApplyMonitorExecutor.schedule(this::scheduleSnapshotApplyVerification, SCHEDULE_APPLY_MONITOR_DELAY,
+                        TimeUnit.MILLISECONDS);
+            }
+        } catch (TimeoutException te) {
+            log.error("Snapshot sync apply verification timed out.", te);
+        } catch (ExecutionException ee) {
+            // Completable future completed exceptionally
+            log.error("Snapshot sync apply verification failed.", ee);
+        } catch (Exception e) {
+            log.error("Snapshot sync apply verification failed.", e);
+        }
```

Comment:
If there is an exception, looks like we are not able to input an event to FSM and not able to start the executor?

Reply:
mmm, you're right we should re-schedule also on any exception: 
this.snapshotSyncApplyMonitorExecutor.schedule(this::scheduleSnapshotApplyVerification, SCHEDULE_APPLY_MONITOR_DELAY,
                        TimeUnit.MILLISECONDS);

I'm not sure if on an unknown exception we should start snapshot sync again.

Reply:
I'm writing a test to be sure we cover these cases.